File: 3.Linked Lists/0.implementation.js
class LinkedList {
    constructor(value) {
        this.head = {
            value: value,
            next: null
        }
        this.tail = this.head;
        this.length = 1;
    }

    append(value){
        const node = new Node(value);
        this.tail.next = node;
        this.tail = node;
        this.length++;
        return this;
    }

    prepend(value){
        const node = new Node(value);
        node.next = this.head;
        this.head = node;
        this.length++;
        return this;
    }

    traverseToIndex(index){
        let counter = 0;
        let current = this.head;
        while(counter !== index){
            current = current.next;
            counter++;
        }
        return current;
    }

    insert(index, value){
        if(index >= this.length){
            return this.append(value);
        }
        if (index === 0){
            return this.prepend(value);
        }
        const node = new Node(value);
        const current = this.traverseToIndex(index - 1);
        node.next = current.next;
        current.next = node;
        this.length++;
        return this;
    }

    remove(index){
        if(index >= this.length){
            throw new Error('Index out of bounds');
        }
        if (index === 0){
            this.head = this.head.next;
            this.length--;
            return this;
        }
        const current = this.traverseToIndex(index - 1);
        let toremove = current.next;
        current.next = toremove.next;
        this.length--;
    }

    printList(){
        const array = [];
        let currentNode = this.head;
        while(currentNode !== null){
            array.push(currentNode.value);
            currentNode = currentNode.next;
        }
        return array;
    }
}

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

let myLinkedList = new LinkedList(10);
myLinkedList.append(5);
myLinkedList.append(16);myLinkedList.prepend(1);
myLinkedList.insert(2, 99);
myLinkedList.insert(20, 88);
myLinkedList.remove(2);

class DoublyLinkedList {
    constructor(value) {
        this.head = {
            value: value,
            next: null,
            prev: null
        }
        this.tail = this.head;
        this.length = 1;
    }

    printList(){
        const array = [];
        let currentNode = this.head;
        while(currentNode !== null){
            array.push(currentNode.value);
            currentNode = currentNode.next;
        }
        return array;
    }

    traverseToIndex(index){
        let counter = 0;
        let current = this.head;
        while(counter !== index){
            current = current.next;
            counter++;
        }
        return current;
    }

    append(value){
        const node = new DoubleNode(value);
        this.tail.next = node;
        node.prev = this.tail;
        this.tail = node;
        this.length++;
        return this;
    }

    prepend(value){
        const node = new DoubleNode(value);
        node.next = this.head;
        this.head.prev = node;
        this.head = node;
        this.length++;
        return this;
    }

    insert(index, value){
        if(index >= this.length){
            return this.append(value);
        }
        if (index === 0){
            return this.prepend(value);
        }
        const node = new DoubleNode(value);
        const current = this.traverseToIndex(index - 1);
        const follower = current.next;
        node.next = follower;
        node.prev = current;
        follower.prev = node;
        current.next = node;
        this.length++;
        return this;
    }

    remove(index){
        if(index >= this.length){
            throw new Error('Index out of bounds');
        }
        if (index === 0){
            this.head = this.head.next;
            this.length--;
            return this;
        }
        const current = this.traverseToIndex(index - 1);
        let toremove = current.next;
        current.next = toremove.next;
        current.next.prev = current;
        this.length--;
    }

    reverse(){
        if(!this.head.next){
            return this.head;
        }
        let first = this.head;
        this.tail = this.head;
        let second = first.next;
        while(second){
            const third = second.next;
            second.next = first;
            first = second;
            second = third;
        }
        this.head.next = null;
        this.head = first;
        return this;
    }
}

class DoubleNode {
    constructor(value) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

File: 3.Linked Lists/0.implementation.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self, value):
        self.head = {
            'value': value,
            'next': None
        }
        self.tail = self.head
        self.length = 1

    def print_list(self):
        array = []
        current_node = self.head
        while current_node != None:
            array.append(current_node.value) # type: ignore
            current_node = current_node.next # type: ignore
        return array
    
    def traverse_to_node(self, index):
        counter = 0
        current = self.head
        while counter != index:
            current = current.next # type: ignore
            counter += 1
        return current
    
    def append(self, value):
        node = Node(value)
        self.tail['next'] = node # type: ignore
        self.tail = node
        self.length += 1
        return self
    
    def prepend(self, value):
        node = Node(value)
        node.next = self.head # type: ignore
        self.head = node
        self.length += 1
        return self
    
    def insert(self, index, value):
        if index == 0:
            return self.prepend(value)
        if index >= self.length:
            return self.append(value)
        node = Node(value)
        current = self.traverse_to_node(index - 1)
        node.next = current.next # type: ignore
        current.next = node # type: ignore
        self.length += 1
        return self
    
    def remove(self, index):
        if index == 0:
            self.head = self.head.next # type: ignore
            self.length -= 1
            return self
        if index >= self.length:
            return self
        current = self.traverse_to_node(index - 1)
        current.next = current.next.next # type: ignore
        self.length -= 1
        return self
    
    def reverse(self):
        if self.head.next == None: # type: ignore
            return self.head
        first = self.head
        self.tail = self.head # type: ignore
        second = first.next # type: ignore
        while second != None:
            third = second.next
            second.next = first # type: ignore
            first = second
            second = third
        self.head.next = None # type: ignore
        self.head = first
        return self

my_linked_list = LinkedList(10)
my_linked_list.append(5)
my_linked_list.append(16)
my_linked_list.prepend(1)
my_linked_list.insert(2, 99)
my_linked_list.insert(20, 88)
my_linked_list.remove(2)
print(my_linked_list.print_list())

class NodeDouble:
    def __init__(self, value):
        self.value = value
        self.next = None

class DoublyLinkedList:
    def __init__(self, value):
        self.head = {
            'value': value,
            'next': None,
            'prev': None
        }
        self.tail = self.head
        self.length = 1

    def print_list(self):
        array = []
        current_node = self.head
        while current_node != None:
            array.append(current_node.value) # type: ignore
            current_node = current_node.next # type: ignore
        return array
    
    def traverse_to_node(self, index):
        counter = 0
        current = self.head
        while counter != index:
            current = current.next # type: ignore
            counter += 1
        return current
    
    def append(self, value):
        node = Node(value)
        self.tail.next = node # type: ignore
        node.prev = self.tail # type: ignore
        self.tail = node
        self.length += 1
        return self
    
    def prepend(self, value):
        node = Node(value)
        node.next = self.head # type: ignore
        self.head.prev = node # type: ignore
        self.head = node
        self.length += 1
        return self
    
    def insert(self, index, value):
        if index == 0:
            return self.prepend(value)
        if index >= self.length:
            return self.append(value)
        node = Node(value)
        current = self.traverse_to_node(index - 1)
        node.next = current.next # type: ignore
        node.prev = current # type: ignore
        node.next.prev = node # type: ignore
        current.next = node # type: ignore
        self.length += 1
        return self
    
    def remove(self, index):
        if index == 0:
            self.head = self.head.next # type: ignore
            self.length -= 1
            return self
        if index >= self.length:
            return self
        current = self.traverse_to_node(index - 1)
        current.next = current.next.next # type: ignore
        current.next.prev = current # type: ignore
        self.length -= 1
        return self
    